enclave/sgx/attester.c

```diff
@@ -145,88 +67,198 @@ static oe_result_t _get_evidence(
-            sizeof(hash.buf),
-            opt_params,
-            opt_params_size,
-            &report,
-            &report_size),
-        "SGX Plugin: Failed to get OE report. %s",
-        oe_result_str(result));
-
-    // Combine the two to get the evidence.
-    // Format is report first then claims.
-    evidence = (uint8_t*)oe_malloc(report_size + claims_size);
-    if (evidence == NULL)
-        OE_RAISE(OE_OUT_OF_MEMORY);
+        if (!memcmp(format_id, &_ecdsa_uuid, sizeof(oe_uuid_t)))
+            format_type = SGX_FORMAT_TYPE_REMOTE;
+        else if (!memcmp(format_id, &_ecdsa_report_uuid, sizeof(oe_uuid_t)))
+            format_type = SGX_FORMAT_TYPE_REMOTE_REPORT;
+        else if (
+            !memcmp(format_id, &_ecdsa_quote_uuid, sizeof(oe_uuid_t)) ||
+            !memcmp(format_id, &_epid_linkable_uuid, sizeof(oe_uuid_t)) ||
+            !memcmp(format_id, &_epid_unlinkable_uuid, sizeof(oe_uuid_t)))
+            format_type = SGX_FORMAT_TYPE_REMOTE_QUOTE;
+        else
+            OE_RAISE(OE_INVALID_PARAMETER);
+    }
 
-    memcpy(evidence, report, report_size);
-    memcpy(evidence + report_size, claims, claims_size);
+    if (format_type == SGX_FORMAT_TYPE_LOCAL ||
+        format_type == SGX_FORMAT_TYPE_REMOTE)
+    { // Evidence of these types has header oe_attestation_header_t
+        OE_SHA256 hash;
 
-    // Get the endorsements from the report if needed.
-    if (endorsements_buffer && flags == OE_REPORT_FLAGS_REMOTE_ATTESTATION)
-    {
-        oe_report_header_t* header = (oe_report_header_t*)report;
+        // Hash the custom_claims.
+        OE_CHECK_MSG(
+            oe_sgx_hash_custom_claims(custom_claims, custom_claims_size, &hash),
+            "SGX Plugin: Failed to hash custom_claims. %s",
+            oe_result_str(result));
 
+        // Get the report with the hash of the custom_claims as the report data.
         OE_CHECK_MSG(
-            oe_get_sgx_endorsements(
-                header->report,
-                header->report_size,
-                &endorsements,
-                &endorsements_size),
-            "SGX Plugin: Failed to get endorsements: %s",
+            oe_get_report_v2_internal(
+                flags,
+                format_id,
+                hash.buf,
+                sizeof(hash.buf),
+                opt_params,
+                opt_params_size,
+                &report,
+                &report_size),
+            "SGX Plugin: Failed to get OE report. %s",
             oe_result_str(result));
+
+        // Combine the report and custom_claims to get the evidence.
+        // Format is attestation header || report || custom_claims.
+        tmp_buffer_size =
+            sizeof(oe_attestation_header_t) + report_size + custom_claims_size;
+        tmp_buffer = (uint8_t*)oe_malloc(tmp_buffer_size);
+        if (tmp_buffer == NULL)
+            OE_RAISE(OE_OUT_OF_MEMORY);
+
+        // Fill the evidence header
+        OE_CHECK(oe_fill_attestation_header(
```

Comment:
So user will have two ways to generate a standard enveloped evidence: `oe_fill_attestation_header()` and `oe_wrap_with_attestation_header()` right?

Reply:
Yes. `oe_wrap_with_attestation_header()` has the overhead to allocate a new buffer for the header plus the data. `oe_fill_attestation_header()` allows a plugin to avoid this overhead if its existing buffer has reserved space for the header.